examples/advanced_operations/add_smart_display_ad.py

```diff
@@ -0,0 +1,267 @@
+    # uploaded should be unique, and should not match another ACTIVE asset in
+    # this customer account.
+    # asset.name.value = f'Jupiter Trip #{uuid4()}'
+    asset_type_enum = client.get_type('AssetTypeEnum', version='v2')
+    asset.type = asset_type_enum.IMAGE
+    image_asset = asset.image_asset
+    image_asset.data.value = image_content
+    image_asset.file_size.value = len(image_content)
+    image_asset.mime_type = client.get_type('MimeTypeEnum').IMAGE_JPEG
+    image_asset.full_size.width_pixels.value = image_width
+    image_asset.full_size.height_pixels.value = image_height
+    image_asset.full_size.url.value = image_url
+
+    asset_service = client.get_service('AssetService', version='v2')
+
+    try:
+        mutate_asset_response = (
+            asset_service.mutate_assets(customer_id, [asset_operation]))
+        return mutate_asset_response.results[0].resource_name
+    except google.ads.google_ads.errors.GoogleAdsException as ex:
+        print(f'Request with ID "{ex.request_id}" failed with status '
+              f'"{ex.error.code().name}" and includes the following errors:')
+        for error in ex.failure.errors:
+            print(f'\tError with message "{error.message}".')
+            if error.location:
+                for field_path_element in error.location.field_path_elements:
+                    print(f'\t\tOn field: {field_path_element.field_name}')
+        sys.exit(1)
+
+
+def _create_responsive_display_ad(client, customer_id, ad_group_resource_name,
+                                  marketing_image_asset_resource_name,
+                                  square_marketing_image_asset_resource_name):
+    ad_group_ad_operation = client.get_type('AdGroupAdOperation', version='v2')
+    ad_group_ad = ad_group_ad_operation.create
+    ad_group_ad.ad_group.value = ad_group_resource_name
+    ad_group_ad.status = client.get_type(
+        'AdGroupAdStatusEnum', version='v2').PAUSED
+    ad = ad_group_ad.ad
+    final_url = ad.final_urls.add()
+    final_url.value = 'https://www.example.com'
+    responsive_display_ad = ad.responsive_display_ad
+    headline = responsive_display_ad.headlines.add()
+    headline.text.value = 'Travel'
+    responsive_display_ad.long_headline.text.value = 'Travel the World'
+    description = responsive_display_ad.descriptions.add()
+    description.text.value = 'Take to the air!'
+    responsive_display_ad.business_name.value = 'Google'
+    marketing_image = responsive_display_ad.marketing_images.add()
+    marketing_image.asset.value = marketing_image_asset_resource_name
+    square_marketing_image = responsive_display_ad.square_marketing_images.add()
+    square_marketing_image.asset.value = (
+        square_marketing_image_asset_resource_name)
+    responsive_display_ad.call_to_action_text.value = 'Shop Now'
+    responsive_display_ad.price_prefix.value = 'as low as'
+    responsive_display_ad.promo_text.value = 'Free shipping!'
```

Comment:
General comment: I prefer not having temporary variables but rather showing the hierarchy of the message. YMMV. I think it is fine to show it both ways in the samples as it is a style preference.